### PR TITLE
Add animehub.ac

### LINKED ISSUE #123: [Feature Request] please add Animehub as one of the sources.
**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by uttiya10**
Note for AnimeHub+Episode:
This PR contains two versions of AnimeHub+Episode.
The "old" version, has enhanced errors as it provides Alternative Server's, if the user's selected server doesn't exist.
However, the "old" version sends two network request (1 -> main website, 1 -> api endpoint). This results in very high load times.

The "new" version, only sends 1 request to the api endpoint. If the user's selected server isn't available, it doesn't provide Alternative Server's. This version is significantly faster.

Please decide whether or not the enhanced errors are worth the high load times.

Close #123

### CODE REVIEW

File: NineAnimator/Models/Anime Source/animehub.ac/AnimeHub+Episode(OLD).swift [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+            // Request api for episode iframe (if available)
+            return self.requestManager.request(
+                "ajax/anime/load_episodes_v2",
+                handling: .ajax,
+                query: ["s": link.server],
+                parameters: ["episode_id": episodeID]
+            ).responseData.then {
+                responseContent in
+                // Convert response into NSDictionary
+                let responseJSON = try JSONSerialization.jsonObject(with: responseContent, options: []) as! NSDictionary
```
Comment (SuperMarcus) [inline NineAnimator/Models/Anime Source/animehub.ac/AnimeHub+Episode(OLD).swift:92]: Avoid force type casting `as!`. Also, if the response is formatted as JSON, why don't you use a struct, and use `responseDecodable`?

File: NineAnimator/Models/Anime Source/animehub.ac/AnimeHub+Episode(OLD).swift [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+                .tryUnwrap()
+            let episodeID = try formDecode(episodeURLParams).value(at: "ep", type: String.self)
+            
+            // Request api for episode iframe (if available)
+            return self.requestManager.request(
+                "ajax/anime/load_episodes_v2",
+                handling: .ajax,
+                query: ["s": link.server],
+                parameters: ["episode_id": episodeID]
+            ).responseData.then {
```
Comment (SuperMarcus) [inline NineAnimator/Models/Anime Source/animehub.ac/AnimeHub+Episode(OLD).swift:89]: If you are using `thenPromise` for this block, promote the next `then` statement to the parent promise.

```Swift
// Deprecated

/* First Promise */ .thenPromise {
    return /* Second Promise */.then {
        return /* Process second promise results */
    }
}

// Recommended

/* First Promise */ .thenPromise {
    return /* Second Promise */
} .then {
    return /* Process second promise results */
}
```

File: NineAnimator/Models/Anime Source/animehub.ac/AnimeHub+Episode.swift [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+            return episodeID
+        } .thenPromise {
+            episodeID in
+            // Request api for episode iframe
+            self.requestManager.request(
+                "ajax/anime/load_episodes_v2",
+                handling: .ajax,
+                query: ["s": link.server],
+                parameters: ["episode_id": episodeID]
+            ).responseData.then {
```
Comment (SuperMarcus) [inline NineAnimator/Models/Anime Source/animehub.ac/AnimeHub+Episode.swift:52]: If you are using `thenPromise` for this block, promote the next `then` statement to the parent promise.

```Swift
// Deprecated

/* First Promise */ .thenPromise {
    return /* Second Promise */.then {
        return /* Process second promise results */
    }
}

// Recommended

/* First Promise */ .thenPromise {
    return /* Second Promise */
} .then {
    return /* Process second promise results */
}
```

File: NineAnimator/Models/Anime Source/animehub.ac/AnimeHub+Episode.swift [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+                responseContent in
+                // Convert response into NSDictionary
+                let responseJSON = try JSONSerialization.jsonObject(with: responseContent, options: []) as! NSDictionary
```
Comment (SuperMarcus) [inline NineAnimator/Models/Anime Source/animehub.ac/AnimeHub+Episode.swift:55]: Avoid forced type casting `as!`.

File: NineAnimator/Models/Anime Source/animehub.ac/AnimeHub.swift [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    var aliases: [String] { [] }
+    
+    #if canImport(UIKit)
+    var siteLogo: UIImage { #imageLiteral(resourceName: "AnimeHub Site Icon") }
+    #elseif canImport(AppKit)
+    var siteLogo: NSImage { #imageLiteral(resourceName: "AnimeHub Site Icon") }
+    #endif
+
+    var siteDescription: String {
+        "AnimeHub has many servers to choose from. NineAnimator has experimental support for this website!"
```
Comment (SuperMarcus) [inline NineAnimator/Models/Anime Source/animehub.ac/AnimeHub.swift:40]:
```suggestion
        "AnimeHub has many servers to choose from. NineAnimator has experimental support for this website."
```

File: NineAnimator/Models/Anime Source/animehub.ac/AnimeHub+Episode(OLD).swift [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+        "fserver": "Fserver",
+        "fdserver": "FDserver",
+        //"xserver": "Xserver", Seems like this no episodes support this server
+        "oserver": "Oserver",
+        "mpserver": "MPserver",
+        "yuserver": "YUserver"
+        //"hserver": "Hserver" Excluding until we update HydraX parser
+    ]
+    /// Represents the response from AnimeHub episode endpoint
+    struct EpisodeResponse: Decodable {
```
Comment (SuperMarcus) [inline NineAnimator/Models/Anime Source/animehub.ac/AnimeHub+Episode(OLD).swift:40]: If `struct EpisodeResponse` is not used anywhere else, you should mark it as `private` or `fileprivate`.

File: NineAnimator/Models/Anime Source/animehub.ac/AnimeHub+Episode.swift [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+//  GNU General Public License for more details.
+//
+//  You should have received a copy of the GNU General Public License
+//  along with NineAnimator.  If not, see <http://www.gnu.org/licenses/>.
+//
+
+import Foundation
+import SwiftSoup
+
+extension NASourceAnimeHub {
```
Comment (SuperMarcus) [inline NineAnimator/Models/Anime Source/animehub.ac/AnimeHub+Episode.swift:23]: Why are there two `AnimeHub+Episode` files?

File: NineAnimator/Models/Anime Source/animehub.ac/AnimeHub.swift [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+        "AnimeHub has many servers to choose from. NineAnimator has experimental support for this website."
+    }
+    
+    var preferredAnimeNameVariant: KeyPath<ListingAnimeName, String> {
+        \.romaji
+    }
+    
+    override var endpoint: String { "https://animehub.ac" }
+    
+    func search(keyword: String) -> ContentProvider {
```
Comment (SuperMarcus) [inline NineAnimator/Models/Anime Source/animehub.ac/AnimeHub.swift:49]: Move this to `AnimeHub+Search.swift`.

File: NineAnimator/Models/Anime Source/animehub.ac/AnimeHub+Episode.swift [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+                query: ["s": link.server],
+                parameters: ["episode_id": episodeID]
+                ).responseDecodable(type: EpisodeResponse.self)
+        } .then {
+            episodeResponse in
+            // Check if server is available.
+            // Note: The api might incorrectly state that the server is available.
+            guard episodeResponse.status == true else {
+                throw NineAnimatorError
+                    .EpisodeServerNotAvailableError(unavailableEpisode: link)
```
Comment (SuperMarcus) [inline NineAnimator/Models/Anime Source/animehub.ac/AnimeHub+Episode.swift:69]: Consider providing the `alternativeEpisodes` and optionally the `updatedServerMap` parameters so `AnimeViewController` can suggest alternative servers that are actually available. Only implement `alternativeEpisodes` if you can tell what servers are available for this episode.

Example code from `AnimeDao+Episode.swift`: https://github.com/SuperMarcus/NineAnimator/blob/2b2086b1d06a000fad180b48b3aa4dfec0531ca4/NineAnimator/Models/Anime%20Source/animedao.com/AnimeDao%2BEpisode.swift#L79-L98

File: NineAnimator.xcodeproj/project.pbxproj [hunk 10]
```diff
@@ -904,6 +913,16 @@
...
+		BF1924A824E1FB12008164C2 /* YourUploadParser.swift */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.swift; path = YourUploadParser.swift; sourceTree = "<group>"; };
+		BFA0704B24DE596D00E7C8BE /* Array+SafeSubscript.swift */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.swift; path = "Array+SafeSubscript.swift"; sourceTree = "<group>"; };
+		BFAA489424DE0723009BE142 /* AnimeHub.swift */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.swift; path = AnimeHub.swift; sourceTree = "<group>"; };
+		BFAA489624DE0BA6009BE142 /* AnimeHub+Featured.swift */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.swift; path = "AnimeHub+Featured.swift"; sourceTree = "<group>"; };
+		BFBAEF5124DF228B00ADA414 /* AnimeHub+Anime.swift */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.swift; path = "AnimeHub+Anime.swift"; sourceTree = "<group>"; };
+		BFC3082024E4E01800943E38 /* AnimeHub+Episode(OLD).swift */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.swift; path = "AnimeHub+Episode(OLD).swift"; sourceTree = "<group>"; };
```
Comment (SuperMarcus) [inline NineAnimator.xcodeproj/project.pbxproj:922]: Why are there two of this file?

**Comment by uttiya10** [comment]
I have now removed the old AnimeHub+Episode

**Pull request merged**